commit d0e204c2303e73b3aaf51acb564b6f47d3d4089d
Author: Lukas Fittl <lukas@fittl.com>
Date:   Wed Feb 17 22:13:53 2021 -0800

    Mark the memory context method-pointer structs "const".
    
    This is a partial backport of 9fa6f00b1308dd10da4eca2f31ccbfc7b35bb461
    which is important for pg_query to avoid making the memory context
    methods thread-local through its automatic rewriting logic.

diff --git a/src/backend/utils/mmgr/aset.c b/src/backend/utils/mmgr/aset.c
index 00328eaf84..97d4e32604 100644
--- a/src/backend/utils/mmgr/aset.c
+++ b/src/backend/utils/mmgr/aset.c
@@ -216,7 +216,7 @@ static void AllocSetCheck(MemoryContext context);
 /*
  * This is the virtual function table for AllocSet contexts.
  */
-static MemoryContextMethods AllocSetMethods = {
+static const MemoryContextMethods AllocSetMethods = {
 	AllocSetAlloc,
 	AllocSetFree,
 	AllocSetRealloc,
diff --git a/src/backend/utils/mmgr/mcxt.c b/src/backend/utils/mmgr/mcxt.c
index 5d173d7e60..ed2877326f 100644
--- a/src/backend/utils/mmgr/mcxt.c
+++ b/src/backend/utils/mmgr/mcxt.c
@@ -638,7 +638,7 @@ MemoryContextContains(MemoryContext context, void *pointer)
  */
 MemoryContext
 MemoryContextCreate(NodeTag tag, Size size,
-					MemoryContextMethods *methods,
+					const MemoryContextMethods *methods,
 					MemoryContext parent,
 					const char *name)
 {
diff --git a/src/backend/utils/mmgr/slab.c b/src/backend/utils/mmgr/slab.c
index 638af8b7aa..29e44f5144 100644
--- a/src/backend/utils/mmgr/slab.c
+++ b/src/backend/utils/mmgr/slab.c
@@ -137,7 +137,7 @@ static void SlabCheck(MemoryContext context);
 /*
  * This is the virtual function table for Slab contexts.
  */
-static MemoryContextMethods SlabMethods = {
+static const MemoryContextMethods SlabMethods = {
 	SlabAlloc,
 	SlabFree,
 	SlabRealloc,
diff --git a/src/include/nodes/memnodes.h b/src/include/nodes/memnodes.h
index 7a0c6763df..60a30b6b74 100644
--- a/src/include/nodes/memnodes.h
+++ b/src/include/nodes/memnodes.h
@@ -76,7 +76,7 @@ typedef struct MemoryContextData
 	/* these two fields are placed here to minimize alignment wastage: */
 	bool		isReset;		/* T = no space alloced since last reset */
 	bool		allowInCritSection; /* allow palloc in critical section */
-	MemoryContextMethods *methods;	/* virtual function table */
+	const MemoryContextMethods *methods;	/* virtual function table */
 	MemoryContext parent;		/* NULL if no parent (toplevel context) */
 	MemoryContext firstchild;	/* head of linked list of children */
 	MemoryContext prevchild;	/* previous child of same parent */
diff --git a/src/include/utils/memutils.h b/src/include/utils/memutils.h
index 869c59dc85..9659ebbcb3 100644
--- a/src/include/utils/memutils.h
+++ b/src/include/utils/memutils.h
@@ -133,7 +133,7 @@ GetMemoryChunkContext(void *pointer)
  * specific creation routines, and noplace else.
  */
 extern MemoryContext MemoryContextCreate(NodeTag tag, Size size,
-					MemoryContextMethods *methods,
+					const MemoryContextMethods *methods,
 					MemoryContext parent,
 					const char *name);
 
